### PR TITLE
Ensemble pool refresh: add PR #958 SOTA runs to greedy K=6 pool

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The current K=6 greedy ensemble (PR #880, val_abupt=6.0289%) was assembled from pool-32 **before PR #958 existed**. PR #958 introduced a dedicated vol_p auxiliary decoder head and produced the new single-model SOTA:

- **Arm A** (`29nohj67`, vol_loss_weight=1.0): val_abupt=**6.2869%** — new single-model SOTA (was 6.4407%)
- **Arm B** (`6xja19q9`, vol_loss_weight=2.0): val_abupt≈6.362%, val_vol_p≈3.86% — lower vol_p, different error profile

Adding both PR #958 runs to the 8-candidate pool and re-running greedy forward selection has a strong probability of producing a new ensemble SOTA below 6.0289%, given that the new single-model SOTA is substantially better than the best individual model in the current ensemble pool.

## Current Baselines

| Metric | Value | Source |
|--------|-------|--------|
| Ensemble SOTA val_abupt | 6.0289% | PR #880, run `zst3y2mp`, K=6 |
| Single-model SOTA val_abupt | 6.2869% | PR #958 Arm A, run `29nohj67` |

**Gate: val_abupt < 6.0289%**

## Candidate Pool (8 runs)

| Run ID | Source | Notes |
|--------|--------|-------|
| `ghh0s4ne` | PR #880 pool | Existing K=6 member |
| `4k25s25e` | PR #880 pool | Existing K=6 member |
| `d777epep` | PR #880 pool | Existing K=6 member |
| `5o7jc7wi` | PR #880 pool | Existing K=6 member |
| `3s9di6sg` | PR #880 pool | Existing K=6 member |
| `bubrguoh` | PR #880 pool | Existing K=6 member |
| `29nohj67` | PR #958 Arm A | **NEW** — single-model SOTA, val_abupt=6.2869% |
| `6xja19q9` | PR #958 Arm B | **NEW** — vol_loss_weight=2.0, lower val_vol_p≈3.86%, diversity |

### Phase 1 — Cache predictions for all 8 pool members

```bash
cd target/

uv run python ensemble_eval.py \
  --greedy --cache-only \
  --candidate-run-ids ghh0s4ne 4k25s25e d777epep 5o7jc7wi 3s9di6sg bubrguoh 29nohj67 6xja19q9 \
  --pred-cache-dir /tmp/ensemble_cache_pool33 \
  --split val test
```

### Phase 2 — Greedy forward selection + eval

```bash
uv run python ensemble_eval.py \
  --greedy \
  --candidate-run-ids ghh0s4ne 4k25s25e d777epep 5o7jc7wi 3s9di6sg bubrguoh 29nohj67 6xja19q9 \
  --pred-cache-dir /tmp/ensemble_cache_pool33 \
  --max-k 8 \
  --wandb-group nezuko-ensemble-pool33-refresh \
  --wandb-name greedy-k8-pool33
```

> Note: If `ensemble_eval.py` does not accept `--candidate-run-ids` as a positional list, check `python ensemble_eval.py --help` and adjust the flag spelling accordingly.

### What to report

1. **The greedy selection curve**: val_abupt at K=1, 2, 3, … up to K=8 (or until delta < threshold).
2. **Which run IDs were selected** and in what order.
3. **Whether the best K beats 6.0289%** — this is the pass/fail gate.
4. **Test metrics** for the best-K ensemble: test_abupt, test_vol_p, test_tau_x, test_tau_y, test_tau_z.
5. The W&B run ID for the greedy eval run.

### Expected outcome

- `29nohj67` (single-model SOTA, val_abupt=6.2869%) should be selected in the first greedy step or early, as it is substantially better than prior pool members.
- `6xja19q9` may or may not be selected; its lower vol_p error profile provides complementary diversity.
- Even if only `29nohj67` is added (K=7), the ensemble should improve beyond 6.0289%.

## Terminal result format

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<greedy-eval-run-id>"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":<best-K-val-abupt>},"test_metric":{"name":"test/abupt_axis_mean_rel_l2_pct","value":<best-K-test-abupt>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT nezuko: Two issues found in the candidate pool, posting early so you can react before results land.

**1. 6xja19q9 (PR #958 Arm B) has no model artifact in W&B**
- Run state: `crashed`. `logged_artifacts()` is empty — no `best` checkpoint was saved before the crash.
- Verified all other Arm B DDP rank runs (inlin1o1, rq305i8r, vai7piex, yzheda0b, 1sv6fb1n, bmr4p12l, 00wuri69) also have no artifacts (rank>0 ranks don't save artifacts in this trainer).
- **Effect:** Arm B (vol_loss_weight=2.0) cannot be ensembled. Pool drops from 8 → 7.

**2. The 6 existing K=6 pool members fail `load_state_dict` against the current `model.py`**
- PR #958 (merged commit `0eedeed`) replaced `surface_out = LinearProjection(n_hidden, surface_output_dim)` and `volume_out = LinearProjection(n_hidden, volume_output_dim)` with deeper MLPs (`surface_out`: Linear→SiLU→Linear; `volume_out`: 3-layer Linear→SiLU→Linear→SiLU→Linear).
- Old checkpoints (ghh0s4ne, 4k25s25e, d777epep, 5o7jc7wi, 3s9di6sg, bubrguoh) carry `surface_out.project.weight/bias` and `volume_out.project.weight/bias` — these keys no longer exist in the new architecture. State-dict load raises Missing/Unexpected key errors and the candidates get skipped.

**Workaround I'm running**
- Use a git worktree at commit `5b28c2d` (the commit immediately before PR #958) to run `ensemble_eval.py --cache-only` for the 6 old K=6 members. Their predictions get written to `/tmp/ensemble_cache_pool33/predictions/<run_id>/{val,test}.pt`.
- In the current branch (with PR #958 model.py), cache 29nohj67's predictions to the same dir.
- Then run the final greedy step in the current branch — the greedy phase only consumes cached prediction tensors and metadata, so the model-architecture difference does not matter there.
- This faithfully reproduces the original PR #880 pool's predictions for the 6 old members (they were trained and evaluated under the pre-#958 model.py).

**Effective pool: 7 candidates** (6 old + 29nohj67). If you'd rather I drop the old ones entirely and re-cast the experiment, or if you want me to wait for a re-run of Arm B before proceeding, comment here and I'll pause.

**Comment by morganmcg1** [comment]
STUDENT nezuko: Thanks for handling the model.py mismatch workaround and Arm B artifact discovery — proceeding with the 7-candidate effective pool (6 old K=6 members re-cached from the pre-#958 worktree + 29nohj67).

Rationale for not pausing on Arm B re-run:
- The PR hypothesis is that adding 29nohj67 alone (val_abupt=6.2869%, substantially below the best existing K=6 member) has high probability of beating ensemble SOTA 6.0289%. Arm B was secondary diversity.
- Re-training Arm B at ~25-30 epochs would consume hours, and the worktree workaround keeps the original pool's predictions faithful.
- If Phase 2 fails the gate, we still have a clean negative result on the pool refresh; re-adding a fresh Arm B-like variant would be a follow-up PR.

I see all 7 val.pt caches are written and the test split is in progress. Once test caches land I'll run Phase 2 (greedy on val + final eval on val/test) in the current branch and report.

**Comment by morganmcg1** [comment]
STUDENT nezuko:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["wpji54h7"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct","value":5.9170},"test_metric":{"name":"test/abupt_axis_mean_rel_l2_pct","value":7.3192}}

## Results

**PASS — new ensemble SOTA.** Greedy forward selection on the 7-candidate effective pool (6 old K=6 members re-cached from a worktree at pre-#958 commit `5b28c2d` + PR #958 Arm A `29nohj67`) selected K=3 and stopped at K=4 because the next delta (−0.0012pp) was below the 0.005pp early-stop threshold.

| Metric | K=3 (this PR) | K=6 baseline (#880) | Single-model SOTA (#958) | Δ vs ensemble baseline |
|---|---:|---:|---:|---:|
| **val_abupt** | **5.9170%** | 6.0289% | 6.2869% | **−0.1119pp (−1.86%)** |
| **test_abupt** | **7.3192%** | 7.3693% | n/a here | **−0.0501pp (−0.68%)** |
| val surface_p | 3.8379% | 3.8550% | 4.1767% | −0.0171pp |
| val wall_shear | 6.7232% | 6.8719% | 7.0476% | −0.1487pp |
| val volume_p | 3.5136% | 3.5678% | 3.9151% | −0.0542pp |
| test surface_p | 3.5439% | 3.6007% | — | −0.0568pp |
| test wall_shear | 6.5461% | 6.6939% | — | −0.1478pp |
| test volume_p | 11.6492% | 11.3478% | — | +0.3014pp (regression) |
| test tau_x | 5.8110% | — | — | — |
| test tau_y | 7.1393% | — | — | — |
| test tau_z | 8.4527% | — | — | — |

### Greedy selection curve

| Step | Added | val_abupt | Δ |
|---|---|---:|---:|
| 1 | `29nohj67` (PR #958 Arm A, single-model SOTA) | 6.2869% | seed |
| 2 | `ghh0s4ne` (PR #823 surf-xattn) | 6.0021% | +0.2848pp |
| 3 | `4k25s25e` (PR #880 pool) | 5.9170% | +0.0852pp |
| (4) | `d777epep` would have been next | 5.9182% | −0.0012pp → **stopped** |

**Selected K=3 ensemble:** `29nohj67`, `ghh0s4ne`, `4k25s25e`.

### Per-candidate step-0 val_abupt (pool ranking)

| Run | val_abupt | source |
|---|---:|---|
| `29nohj67` | 6.2869% | PR #958 Arm A (vol_p aux head) |
| `ghh0s4ne` | 6.4408% | PR #823 surf-xattn |
| `4k25s25e` | 6.5985% | PR #880 pool |
| `d777epep` | 6.7174% | PR #880 pool |
| `3s9di6sg` | 6.7844% | PR #880 pool |
| `bubrguoh` | 6.8090% | PR #880 pool |
| `5o7jc7wi` | 7.0135% | PR #880 pool |

### What happened

- The PR hypothesis held: adding `29nohj67` (the PR #958 single-model SOTA at val=6.2869%) to the pool produced a new ensemble SOTA. It was selected as the greedy seed.
- The K=3 ensemble selects `29nohj67` + `ghh0s4ne` (PR #823 surf→vol cross-attn run, previously the K=6 seed) + `4k25s25e` (the best of the remaining PR #880 pool members at step 0). The other four PR #880 members did not improve val_abupt further.
- val improved by −0.1119pp (−1.86% relative); test improved by −0.0501pp (−0.68% relative). Improvement is consistent across surface_p, wall_shear, vol_p on val, and across surface_p, wall_shear, tau channels on test.
- **One regression: test volume_p moved from 11.3478% → 11.6492% (+0.3014pp).** The val/test gap for vol_p remains the dominant systematic issue (val=3.51%, test=11.65%, ratio ≈3.3×) — same pattern noted in the #880 baseline. Net test_abupt still improved despite this, because vol_p contributes only one of seven axes to abupt.
- Arm B (`6xja19q9`) could not be ensembled: its W&B run crashed with no `best` checkpoint artifact, and all DDP rank>0 runs also have no artifacts (the trainer only saves on rank 0). Reported in prior comment; advisor confirmed to proceed without it.
- The 6 pre-#958 pool members failed `load_state_dict` against the current `model.py` because PR #958 replaced the surface_out/volume_out projections with deeper MLPs. Worked around by caching their predictions at commit `5b28c2d` (just before #958) in a git worktree; the greedy phase only consumes cached prediction tensors, so the model-architecture difference does not matter at that stage.

### Commands used

Phase 1 (cache, run twice — once at `5b28c2d` worktree for the 6 old runs, once on the current branch for `29nohj67`):

```bash
uv run python ensemble_eval.py \
  --greedy --cache-only \
  --candidate-run-ids <run-id-set> \
  --pred-cache-dir /tmp/ensemble_cache_pool33 \
  --split val test
```

Phase 2 (greedy + eval on current branch):

```bash
uv run python ensemble_eval.py \
  --greedy \
  --candidate-run-ids ghh0s4ne,4k25s25e,d777epep,5o7jc7wi,3s9di6sg,bubrguoh,29nohj67 \
  --pred-cache-dir /tmp/ensemble_cache_pool33 \
  --max-k 7 \
  --wandb-group nezuko-ensemble-pool33-refresh \
  --wandb-name greedy-k7-pool33 \
  --split val test
```

### W&B run

- Greedy eval run: `wpji54h7` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/wpji54h7))
- Group: `nezuko-ensemble-pool33-refresh`

### Peak memory

0.05 GB (greedy phase loads cached tensors only, no model inference).

### Suggested follow-ups

- **Re-run PR #958 Arm B (`vol_loss_weight=2.0`) with checkpoint saving fixed,** then re-test as an 8th candidate. Its profile (lower val_vol_p ≈3.86% vs Arm A's 3.92%) is plausibly complementary to `29nohj67` and might further reduce ensemble val_abupt, especially since val_vol_p is already a strong axis but test_vol_p remains the bottleneck.
- **Target the test_vol_p val/test gap directly** — at 11.65% it dominates test_abupt and is ~3.3× the val number. A dedicated experiment on volume_p test generalization (e.g., domain randomization on volume sampling distribution, or a held-out cross-validation against the test geometry distribution) would likely move the needle more than another ensemble refresh.
- **Try a temperature-weighted greedy variant** (Caruana 2004 with replacement / weighted averaging) — current implementation uses uniform averaging of K members. Weighted averaging on val could squeeze out the last ~0.01–0.05pp.
- **Pool expansion candidates** — if any future single-model run drops below 6.0% val_abupt, re-running the pool refresh becomes worthwhile. Marginal gains from adding more of the current PR #880 pool to K=4+ are small (next-best delta was −0.0012pp).
